### ISSUE TITLE
list blobs with marker

### ISSUE THREAD

**Issue by SergeVil** [Azure/azure-storage-java#128]
Hi All,

I saw that such functionality exists, but what is the Java API for it?

Thank you!
Serge

**Comment by jofriedm-msft** [comment]
Hi @SergeVil,

 listBlobsSegmented() has an overload with the parameter "ResultContinuation continuationToken".
The ResultContinuation object has a "String nextMarker" field.

Let us know if you have any further questions.

**Comment by SergeVil** [comment]
Hi ,

My question is about how to provide the marker to the request and not how to read the next marker from the response (its clear from the API). What I want to do is for a very large amount of blobs within the container, to have a list of markers and run several threads that each one will bring the portion of the blobs names starting from the particular marker. That will let me read such a large container in parallel, which is much faster. I see that the REST document says that there is such an option, https://docs.microsoft.com/en-us/rest/api/storageservices/fileservices/List-Blobs?redirectedfrom=MSDN, but I do not see it in Java API.

Thanks,
Serge

**Comment by jofriedm-msft** [comment]
Hi @SergeVil ,

That API is for setting the marker. The full signature is

@param continuationToken
     *            A {@link ResultContinuation} object that represents a continuation token returned
     *            by a previous listing operation.
public ResultSegment<ListBlobItem> listBlobsSegmented(final String prefix, final boolean useFlatBlobListing,
            final EnumSet<BlobListingDetails> listingDetails, final Integer maxResults,
            **final ResultContinuation continuationToken,** BlobRequestOptions options, OperationContext opContext)

The ResultContinuationToken contains 

/**
     * Sets the next marker for continuing results of listing operations.
     * 
     * @param nextMarker
     *            A <code>String</code> which represents the the next marker to set.
     */
    public void setNextMarker(final String nextMarker) {
        this.nextMarker = nextMarker;
    }

Markers for Azure cannot be generated by a client. This is the API to pass in a marker but the marker will need to be generated by the service. You can also consider setting a prefix and listing in parallel starting at different locations alphabetically..

**Comment by SergeVil** [comment]
Hi Josh,

I see. What is the best practice to parallelize the full list reading? Setting the prefix is good only if the files within directories are proportionally distributed. 

Is it anything like http://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/s3/model/ListObjectsV2Request.html#setStartAfter-java.lang.String- ?

Thanks,
Serge

**Comment by jofriedm-msft** [comment]
Hi @SergeVil ,

Best practice is to use a prefix. If you have a notion of how your blobs are distributed alphabetically or have longer prefixes you can use besides a single letter, this can help with listing in parallel.

**Comment by SergeVil** [comment]
If all millions files are within the same directory, how prefixes could help? setStart would help as I can read it once and make some marks that in the future read fragments starting from these marks in parallel.

Serge

**Comment by jofriedm-msft** [comment]
Hi @SergeVil, If several files all start with the same prefix ie "ab1, ab2, ab3, cd1, cd2" you could use "ab" and "cd" as a prefix.

Another option you could do is save the continuation tokens that you received from the first time listing. Then for future listing operations, the listing can be done in parallel using the different continuation tokens. Keeping in mind that results may not be accurate if blobs have not been added and delete.

**Comment by SergeVil** [comment]
I would be nice to have the option of setStart. If you work for Microsoft, please, take it into consideration. Otherwise its hard to manage big catalogs fast. 

Serge

**Comment by jofriedm-msft** [comment]
We will take this as a feature request and look into making listing more efficient. Please let us know if you have another issues.
